app/models/campaign_page.rb

```diff
@@ -7,16 +7,25 @@ class CampaignPage < ActiveRecord::Base
   has_many :campaign_pages_widgets, inverse_of: :campaign_page
   has_many :campaign_pages_tags
   has_many :tags, through: :campaign_pages_tags, source: :tag
+  has_many :widgets, dependent: :destroy
```

Comment:
this part is very much a work in progress. I'm planning to delete the CampaignPagesWidgets class,  but I left it in place as I was kinda just sprinting to show off what I was thinking.